File: tests/cases/arr_repeat_check_panic.ts
// A repeat whose check was kept still panics with the index message: each of
// these reads its index once in range and once past the end, after a `pop`, a
// call that rebinds the field, and a field store. `arr_repeat_check_panic.c`
// runs each in a child process and prints what it wrote and how it exited.
const afterPop = (xs: i32[], i: i32): i32 => {
  const a = xs[i];
  xs.pop();
  return a + xs[i];
};

class Box {
  v: i32[];

  constructor() {
    this.v = [1, 2, 3, 4];
  }

  shrink(): i32 {
    this.v = [5];
    return 0;
  }

  afterCallStore(i: i32): i32 {
    this.v[i] = this.shrink();
    return this.v[i];
  }

  afterFieldStore(i: i32): i32 {
    const a = this.v[i];
    this.v = [6];
    return a + this.v[i];
  }
}

export const popPastEnd = (): i32 => afterPop([1, 2, 3], 2);

export const callStorePastEnd = (): i32 => new Box().afterCallStore(3);

export const fieldStorePastEnd = (): i32 => new Box().afterFieldStore(3);

export const test = (): number => afterPop([1, 2, 3], 1) + new Box().afterFieldStore(0);

File: tests/cases/arr_repeat_check_panic.c
/* Driver for arr_repeat_check_panic: `test()` in range, then each out-of-range
   repeat in a child whose stderr is this process's stdout, so the panic
   message and the exit status land in the .out file. */
#include <stdint.h>
#include <stdio.h>
#include <sys/wait.h>
#include <unistd.h>

int32_t test(void);
int32_t popPastEnd(void);
int32_t callStorePastEnd(void);
int32_t fieldStorePastEnd(void);

static void expectPanic(const char *name, int32_t (*run)(void)) {
  fflush(stdout);
  pid_t pid = fork();
  if (pid == 0) {
    dup2(1, 2);
    printf("%s returned %d\n", name, run());
    fflush(stdout);
    _exit(0);
  }
  int status = 0;
  waitpid(pid, &status, 0);
  printf("%s exit %d\n", name, WIFEXITED(status) ? WEXITSTATUS(status) : -1);
}

int main(void) {
  printf("%d\n", test());
  expectPanic("popPastEnd", popPastEnd);
  expectPanic("callStorePastEnd", callStorePastEnd);
  expectPanic("fieldStorePastEnd", fieldStorePastEnd);
  return 0;
}
